docs: parachain onboarding locally (#10)
r0gue-io/pop-docs: pop-cli-for-appchains/guides/running-on-paseo-locally.md

---
description: >-
  This guide shows how to manually onboard a chain to Paseo Local
---

# Onboard a Chain to Paseo Local

[Paseo](https://x.com/PaseoNetwork) is the community-run Polkadot Relay chain Testnet. Paseo Local is for running Paseo on a local machine.

> A typical development workflow for launching a chain on Polkadot:&#x20;

Reviewer: Afraid of adding content to this already long doc. 
But I feel we could use some extra pointers for the bullet points bellow.

#1. Is very useful when testing the runtime of your chain. You are not really worried about how the onboarding process works. But you need to see the runtime executing some blocks. "live" integration tests with other chains and such.

#2. Is only helpful when you are trying to learn how to onboard your chain to an already running relay chain (like Westend, Paseo, Kusama or Polkadot). It is very helpful to nail down the concepts around bootstrapping collators, seeing the process and taking a few leanings from it.
This document.

The learnings in #2. are mostly equivalent to #3. in terms of what the user needs to know.
I'd argue that we can port most of the new content from here into 3. (or vice versa) and get rid of all the details around configuring  the relay chain. It is too nuanced, and possibly it changes in future versions of the SDK.

Ideally no user would need to know about the scheduling gymnastics the relay does.

Author: This was a bit hard to follow what you meant. Please see the latest changes and shout if there is anything pressing missing (or create an issue if not pressing).

  Testnet Locally
---

# Launch a Chain on a Local Test Network

## Introduction

[Paseo](https://x.com/PaseoNetwork) is the community-run Polkadot Relay chain Testnet.

A typical development workflow for launching your chain on Polkadot:&#x20;

* Run your chain **locally** using Pop CLI.&#x20;

>   * Under the hood, Pop CLI launches your chain to Paseo Local Testnet automatically for development purposes:

Reviewer 2: Here it is used `Paseo Local Testnet`

    * [Run your parachain on Paseo](running-your-parachain.md)
* When ready to test your chain in a live environment with other chains:

>   * Use this guide to mimic the manual onboarding process for Paseo Testnet locally.&#x20;

Reviewer 2: ```suggestion
  * Use this guide to mimic the manual onboarding process for a Paseo Local Testnet.&#x20;
```

As we call a production Paseo network `Paseo Live Testnet`, I suggest following the format `Paseo [Live | Local] Testnet` to keep the consistency.

Reviewer 2: I guess this is Paseo Live Testnet?

Author: Yes, great suggestion

Author: I am calling it Paseo Local and Paseo, the latter being live. Also removing the word testnet as that is explained via a link at the top.

  * When comfortable with manually onboarding locally, then use the next guide to onboard to Paseo Live Testnet.

>     * [Launching your parachain on Paseo Testnet](launching-your-parachain-on-polkadot/launching-on-paseo.md)

Reviewer 2: And then here it is called `Paseo Testnet`, a bit confusing to me whether this is Live Testnet or Local Testnet.

  Testnet Locally
---

# Launching your parachain manually
# Launch your parachain on a Local Test Network.

> ## Introduction

Author: You feel it is necessary?

> ## Introduction

Reviewer 3: I don't see it as a bad content a brief explanation on what Paseo is, maybe is too long. But not blocker for me.

> ## Introduction

Author: Moved it back

Let's get started.

> ## Launch Paseo Local

Reviewer 2: ```suggestion
## Launch Paseo Testnet on Local
```

or 

```suggestion
## Launch Paseo Local Testnet
```

validator = true
```

> > Interesting Fact: We have three validator nodes so that we can use Polkadot's Warp Sync to quickly sync with the network. The minimum requirement for Warp Sync are three validator nodes.

Reviewer 3: Again, why remove it?

Author: You feel it is necessary? Someone wants to deploy a parachain, there is already a lot of info.

to Paseo if needed.

Run the network:
```
pop up parachain -f network.toml --verbose
```

> The `--verbose` flag provides us with extra information such as the location of the Paseo chain spec file. This will become important later so keep this terminal tab as is.

<figure><img src="../.gitbook/assets/Screenshot 2024-09-24 at 12.20.38 PM.png" alt=""><figcaption><p>pop up parachain -f network.toml --verbose</p></figcaption></figure>

Paseo should now be running on your machine and producing blocks.

### Configure Paseo

As of now, Paseo local doesn't provide cores to validate parachain blocks on demand. We will have to make 2 calls to Paseo
using `Alice` as admin account.

First, configure Paseo to set coretime cores to `1`:
```bash

> pop call parachain --pallet Configuration --function set_coretime_cores --args "1" --url ws://localhost:57731/ --suri //Alice --sudo --skip-confirm

Reviewer 3: Use `pop call chain`

```

> > Note: we are calling the specified rpc port `577313` which is specified in the created `network.toml` file to interact the validator `alice`.

Reviewer 3: ```suggestion
> Note: we are calling the specified rpc port `57731` which is specified in the created `network.toml` file to interact the validator `alice`
```

Second, assign the core to the on demand pool:
```bash

> pop call parachain --url ws://localhost:57731 --call 0xff004a0400000a000000040100e100 --suri //Alice --skip-confirm

Reviewer 3: Use `pop call chain`

> pop call parachain --url ws://localhost:57731 --call 0xff004a0400000a000000040100e100 --suri //Alice --skip-confirm

Reviewer 3: Why the call_data here? Is not clear what we are doing if we specify the extrinsic we are calling?

<figure><img src="../.gitbook/assets/Screenshot 2024-09-24 at 12.20.38 PM.png" alt=""><figcaption><p>pop up parachain -f network.toml --verbose</p></figcaption></figure>

Paseo should now be running on your machine and producing blocks.

### Configure Paseo

As of now, Paseo local doesn't provide cores to validate parachain blocks on demand. We will have to make 2 calls to Paseo
using `Alice` as admin account.

First, configure Paseo to set coretime cores to `1`:
```bash
pop call chain --pallet Configuration --function set_coretime_cores --args "1" --url ws://localhost:57731/ --suri //Alice --sudo --skip-confirm
```

> Note: the specified rpc port `57731` is specified in the created `network.toml` file and is to interact with the validator `alice`.

Second, assign the core to the on demand pool:
```bash
pop call chain --url ws://localhost:57731 --call 0xff004a0400000a000000040100e100 --suri //Alice --skip-confirm
```

## Setting Up Accounts

First, we will need to set up a stash account to do transactions on Paseo on behalf of our collator that we will launch later in this guide.

> > A collator is the parachain node that will be running for your parachain.

Reviewer 2: ```suggestion
> A [collator](https://wiki.polkadot.network/docs/learn-collator) is the parachain node that will be running for your parachain.
```

Author: Great suggestion, thank you!

> This your key (Polkadot account). Save the secret phrase in a vault securely and never share it.

Now that we have a stash account, we need to fund this account with some tokens so that it has funds to perform transactions on behalf of the collator such as transactions related to onboarding your parachain.

> Add the stash account to the [Polkadot Signer extension](https://polkadot.js.org/) so that you can see your account appear in the [PolkadotJs Apps](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fpaseo-rpc.dwellir.com#/accounts) UI.

Reviewer 3: Why remove this part?

Author: Do we need it? We do everything with pop cli

Reviewer 3: True!

Build your parachain for release:

```
pop build --release
```
### Creating the chain spec

Create the chain spec:
The chain specification holds all the information the node requires to start or sync with the parachain network.

Let's create a chain spec for our parachain:

```
pop build spec --release --id 2000 --type local --relay paseo-local --protocol-id my_parachain
cd my-parachain
pop build spec --profile release --id 2000 --type local --relay paseo-local --protocol-id my_parachain --chain local --genesis-state --genesis-code
```
```bash
┌   Pop CLI : Generate your chain spec
│
◇  Name or path for the plain chain spec file:
│  ./chain-spec.json
│
◇  Would you like to use local host as a bootnode ?
│  No 
│
┌   Pop CLI : Building your chain spec
│
◇  Chain specification built successfully.
│
◆  Generated files:
│  ● Plain text chain specification file generated at: ./chain-spec.json
│  ● Raw chain specification file generated at: ./chain-spec-raw.json
│  ● WebAssembly runtime file exported at: ./para-2000.wasm
│  ● Genesis State file exported at: ./para-2000-genesis-state
│  
└  Need help? Learn more at https://learn.onpop.io
```

> This will output the following:

Reviewer 3: Is not interesting to keep this?

Author: See the bash script saying exactly the same

Reviewer 3: You are right, sorry

> Save the secret phrase in a secure manner.

We now have all the accounts we need and we can start prepare our chain!

## Setting up the Chain

For the sake of this exercise, let's create a new chain project:
```
pop new parachain my-chain
```
> The folder includes a `network.toml` file which can be ignored. This is to launch a network with the chain already onboarded.

### Creating the chain spec

The chain specification holds all the information the node requires to start or sync with the chain's network.

Let's generate a chain spec:

<figure><img src="../.gitbook/assets/buildspec.gif" alt="pop build spec"><figcaption><p>pop build spec</p></figcaption></figure>

```
cd my-chain
pop build spec --profile release --id 2000 --type local --relay paseo-local --protocol-id my_chain --chain local --genesis-state --genesis-code
```

> ```bash

Reviewer 4: Do we still need this output if we have the gif?

Author: For copy pasting it can still be handy perhaps.

Author: You are right, removed the command and output.

* **add your account and session keys**
* **specify the starting balance of specific accounts**

> * **add the account that will be the sudo account for your chain**

Reviewer: I am a bit skeptical about inviting users to modify the plain spec directly instead of having code in the project that gives them the expected genesis.

But I think this is a different conversation than what this PR involves.

Author: Yes out of scope I'm afraid

In order to run your parachain's collator you will need the raw chain spec that Paseo Local is using.

This can be found in the output when you ran the `pop up parachain -f network --verbose` command:

<figure><img src="../.gitbook/assets/Screenshot 2024-09-24 at 12.20.38 PM (1).png" alt=""><figcaption></figcaption></figure>

Copy this chain spec into our `my-chain` directory:

```bash
cd my-chain
cp /var/folders/vl/txnq6gdj22s9rn296z0md27w0000gn/T/zombie-ddb5d2aa-704b-4658-af64-3cf9e3be5573/alice/cfg/paseo-local.json paseo-local-raw.json
```

> Note: Your Paseo Local chain spec path will be different from the above.

### Generate Node Key

We will also need to create a node-key for your collator:

```bash
cd my-chain
mkdir -p data/chains/my_chain/network
docker run -it parity/subkey:latest generate-node-key > ./data/chains/my_chain/network/secret_ed25519
```

> Alternatively you can use Polkadot SDK binary instead of a Docker image:
> ```
> path/to/polkadot-sdk/target/debug/substrate-node key generate-node-key --file=secret_ed25519 --chain=./chain-spec-raw.json
> ```
>
> <pre><code><strong>mv secret_ed25519 data/chains/my_chain/network
> </strong></code></pre>

> ### Run Collator

Reviewer: Note that the above node key is placed in the `.data/` directory assuming that the user will run the collator copy/pasting the instructions.

More concretely the node key generated in the previous step will only be used if `--base-path ./data ` is provided.

Authorities need to be run with a specific node key. Otherwise `--unsafe-force-node-key-generation` can be provided to delegate the key generation to the node itself (this presents other nuances). But this is OK for local environments.

Author: Added a little `Note`

http://localhost:8845
```

Ensure the chain is synced with Paseo:
```bash
2024-12-10 09:06:05 [Relaychain] Warp sync is complete, continuing with state sync.    
2024-12-10 09:06:06 [Relaychain] State sync is complete, continuing with block sync.    
2024-12-10 09:06:06 [Relaychain] 🏆 Imported #47 (0xb04f…12a4 → 0x6adf…6582)    
2024-12-10 09:06:06 [Relaychain] 🏆 Imported #48 (0x6adf…6582 → 0xd036…e502) 
```

We now need to onboard the chain to Paseo.

### Reserve Para ID

We can reserve a para ID for the chain using pop cli:

```bash
pop call parachain --url ws://localhost:57731
```

 
```bash
┌   Pop CLI : Call a parachain
│
◇  What would you like to do?
│  Reserve a parachain ID 
│
◇  Signer of the extrinsic:

> │  <PUBLIC KEY STASH ACCOUNT>

Reviewer 3: That's wrong no?

Author: Why?

Reviewer 3: Here is the Secret key URI, not the public. maybe just leave it as `<STASH ACCOUNT>`

│  
◇  Do you want to submit the extrinsic?
│  Yes 
│

Event(s):

> Registrar::Reserved: { para_id: (2000), who: ((212, 53, 147, 199, 21, 253, 211, 28, 97, 20, 26, 189, 4, 169, 159, 214, 130, 44, 133, 88, 133, 76, 205, 227, 154, 86, 132, 231, 165, 109, 162, 125)) }

Reviewer 3: Output is outdated.

Author: Updated

http://localhost:8845
```

Ensure the chain is synced with Paseo:
```bash
2024-12-10 09:06:05 [Relaychain] Warp sync is complete, continuing with state sync.    
2024-12-10 09:06:06 [Relaychain] State sync is complete, continuing with block sync.    
2024-12-10 09:06:06 [Relaychain] 🏆 Imported #47 (0xb04f…12a4 → 0x6adf…6582)    
2024-12-10 09:06:06 [Relaychain] 🏆 Imported #48 (0x6adf…6582 → 0xd036…e502) 
```

We now need to onboard the chain to Paseo.

### Reserve Para ID

We can reserve a para ID for the chain using pop cli:

```bash
pop call chain --url ws://localhost:57731
```

 
```bash
┌   Pop CLI : Call a chain
│
◇  What would you like to do?
│  Reserve a parachain ID 
│

> ◇  Signer of the extrinsic:

Reviewer: We could include `--use-wallet` to exemplify a more solid flow.
Otherwise we can come back once the feature its merged!

│  <STASH ACCOUNT>
│  
...
       Event Balances ➜ Reserved
         who: <STASH ACCOUNT>
         amount: 100UNIT
       Event Registrar ➜ Reserved
         para_id: Id(2000)
         who: <STASH ACCOUNT>
...         
```

In the events we can see the `para_id` that is assigned to the chain. Make sure this is the para ID specified in the chain spec file (and thus the chain artifacts).

### Register Para ID with Genesis State & Code

Now we register the para ID with the generated genesis state (`para-2000-genesis-state`) and genesis code (`para-2000.wasm`).

```bash
│
◇  Do you want to perform another call?
│  Yes
│
◇  What would you like to do?
│  Register a parachain ID with genesis state and code
│
◇  Enter the value for the parameter: id
│  2000
│
◇  The value for `genesis_head` might be too large to enter. You may enter the path to a file instead.
│  para-2000-genesis-state
│
◇  The value for `validation_code` might be too large to enter. You may enter the path to a file instead.
│  para-2000.wasm
│

> ◇  Signer of the extrinsic:

Reviewer: Same comment than above regarding `--use-wallet`.

│  <STASH ACCOUNT>
│
...
       Event Balances ➜ Withdraw
         who: <STASH ACCOUNT>
         amount: 90.71989507390UNIT
       Event Balances ➜ Reserved
         who: <STASH ACCOUNT>
         amount: 3.145826kUNIT
       Event Paras ➜ PvfCheckStarted
         0: ValidationCodeHash(0x1821617486094e18595084b580fe9324a084adedbf80ec61c9d7b75736ab5f5b)
         1: Id(2000)
       Event Registrar ➜ Registered
         para_id: Id(2000)
         manager: <STASH ACCOUNT>
...
```

Your chain is now registered on Paseo!

> ### Buy an On Demand Core

Reviewer: Might be interesting pointing to the wiki - https://wiki.polkadot.network/docs/learn-agile-coretime
As a reference in case someone arrives to this point and is not aware of what on demand coretime is. :)

Author: Nice one!

Reviewer 4: 'Buy a core' doesnt sound right to me - we are acquiring coretime for a single block where this sounds like you have bought it outright and own it, which is not true.

https://wiki.polkadot.network/docs/learn-agile-coretime#on-demand-coretime

Author: Changed to acquire a core

> Now we need to buy a core to have Paseo validate a block.

Reviewer: Not 100% correct. One only buy the finalization of one core.
More of a pay finalization as you produce blocks.

Buying a core would be something more akin to buying bulk coretime. Which in reality isn't cores either what is being sold, but regions. :)

Author: validated and finalised*
